{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8f9e90d",
   "metadata": {},
   "source": [
    "# Assessment \n",
    "Answer the following questions with python using Jupyter notebook. You should include all the questions, your code, the outcome of the code and your answers in a single file, \"html\" or \"PDF\". Use the download as option from the file menu to do this. You can attach your *ipynb file as well, but make sure it works with outcomes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a0c8586",
   "metadata": {},
   "source": [
    "# Question 1 [20%]\n",
    "\n",
    "You have a data file called TCGA.csv, which has the gene expression dataset of some patients with liver cancer from the TCGA project.\n",
    "\n",
    "**Using the data, complete the below tasks:**\n",
    "* 1. Create a new data frame containing samples S15, S16, S17, S18, S19, S20 and gene columns.\n",
    "* 2. Using the data in 1. above, add a new column called group by dividing the values of S18 into high and low based on the median value. \n",
    "* 3. Extract the list of genes with values in S20 column lower than 10, and the group in 2 is high.\n",
    "* 4. Create a new data frame from 2. above, with groups equal to low and values less than 3 in S16. \n",
    "* 5. Create a scatter plot for the expression in S1 against S18 (apply log10 transformation to the plot). \n",
    "* 6. Create a boxplot for expression in S19 against the group in data 2. above. Comment on your observation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a9c12e9",
   "metadata": {},
   "source": [
    "# Question 2 [30%]\n",
    "\n",
    "**Write functions that do the following tasks:**\n",
    "\n",
    "* 1. That takes a DNA sequence (for example \"CCTGTCAAAAGATATGT\") and converts it to an RNA sequence. Then returns the DNA sequence, the RNA sequence and the length of this sequence using a list. [Hint: Convert all the T in DNA to U to make RNA. Use the \"string\".repalce() method] [10%]\n",
    "\n",
    "* 2. Write a function that takes the size of a chromosome and creates a non-overlapping window of a size specified by an argument. Therefore, chromosome size 100 and size 20 return 1-20, 21-40, 41-60, 61-80, 81-100. This function should have an optional argument to either print the result on the screen or not. [10%]\n",
    "\n",
    "* 3. Write a function to determine the maximum of four numbers given as a tuple argument. [10%]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78bf82e5",
   "metadata": {},
   "source": [
    "# Question 3 [20%]\n",
    "\n",
    "We have defined a Cell class below. Write a subclass called adTCell that extends the Cell class. It should have self.protein, self.size set to 1, and self.max_rate attributes. The self.protein attribute should be a list.\n",
    "\n",
    "\n",
    "Code the following method:\n",
    "1. .init(*****, size) constructor. Identify and include all the parameters to ensure that the cell .init() is included.\n",
    "2. Set self.size to size and self.max_rate to 0.5\n",
    "3. .add_protein() that takes a protein name as an argument and adds it to the self.protein attribute.\n",
    "4. .summary() that prints the total number of organelle, genes and proteins.\n",
    "5. .grow() that takes a growth rate argument and increases the self.size attribute by that rate if it is smaller than the self.max_rate, otherwise complain and do nothing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6e57a46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is the cell class to use for Question 3 above.\n",
    "\n",
    "class Cell:\n",
    "    \"\"\" A class representing a cell\"\"\"\n",
    "    \n",
    "    # Constructor \n",
    "    def __init__(self, name):\n",
    "        \"\"\"Input layer\"\"\"\n",
    "        self.name = name\n",
    "        self.organelles = []    # creates a new empty list for each cell\n",
    "        self.genes = []         # same as above\n",
    "    \n",
    "    # The string method\n",
    "    def __str__(self):\n",
    "        return (\"A model of a cell, current instance is {} \".format(self.name))\n",
    "    \n",
    "    # Methods\n",
    "    def add_organelle(self, organelle):\n",
    "        self.organelles.append(organelle)\n",
    "    \n",
    "    def add_gene(self, gene):\n",
    "        self.genes.append(gene)\n",
    "        \n",
    "    def view_cell(self):\n",
    "        return self.organelles, self.genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d840ae4",
   "metadata": {},
   "source": [
    "# Question 4 [30%]\n",
    "In your R assessment 3 you were asked to do functions to perform some tasks. Please, convert the R code for the below question to the python code to do the same tasks.\n",
    "\n",
    "Write functions that can do the following:\n",
    "* 1. Take a string containing a DNA sequence, and convert this to a vector, which each entry is one base.\n",
    "  a. Calculate the length of the sequences.\n",
    "  b. The amount of each of the four nucleotides, A, T, G and C. \n",
    "  c. Return these as a tuple with each item named (i.e. length, A, T, G, C)\n",
    "  [15%]\n",
    "  \n",
    "  \n",
    "* 2. Write a function that takes a DNA sequence and substitutes one nucleotide specified by an argument (default value “A”) for another nucleotide specified by an argument (default value “T”, in other words, the default of this function changes “A” to “T”). This function should return a dictionary with both the original and the new string of nucleotides. [15%]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
