{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "235056fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accession code:\n",
      ">P04637\n",
      "\n",
      "Name:\n",
      "P53_HUMAN Cellular tumor antigen p53 - Homo sapiens (Human).\n",
      "\n",
      "Protein:\n",
      "MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD\n",
      "\n",
      "Number of residues:\n",
      "393\n"
     ]
    }
   ],
   "source": [
    "# 1\n",
    "\n",
    "FASTA = open(\"P89u.fas\", \"r\")\n",
    "header = FASTA.readline()\n",
    "\n",
    "protein = \"\" # build up the sequence here\n",
    "\n",
    "for ll in FASTA:\n",
    "    protein += ll.rstrip() # remove trailing '\\n'\n",
    "\n",
    "FASTA.close()\n",
    "\n",
    "# Print the header\n",
    "(code, name) = header.split('|')\n",
    "\n",
    "print(\"Accession code:\")\n",
    "print(code)\n",
    "\n",
    "print(\"\\nName:\")\n",
    "print(name)\n",
    "\n",
    "print(\"Protein:\")\n",
    "print(protein)\n",
    "\n",
    "print(\"\\nNumber of residues:\")\n",
    "print(len(protein))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bab36c4",
   "metadata": {},
   "source": [
    "# 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fbe652ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def print_protein_info(file=\"P89u.fas\"):\n",
    "    FASTA = open(file, \"r\")\n",
    "    header = FASTA.readline()\n",
    "    \n",
    "    protein = \"\" # build up the sequence here\n",
    "    \n",
    "    for ll in FASTA:\n",
    "        protein += ll.rstrip() # remove trailing '\\n'\n",
    "        \n",
    "    FASTA.close()\n",
    "    \n",
    "    # Print the header\n",
    "    (code, name) = header.split('|')\n",
    "    \n",
    "    print(\"Accession code:\")\n",
    "    print(code)\n",
    "\n",
    "    print(\"\\nName:\")\n",
    "    print(name)\n",
    "\n",
    "    print(\"Protein:\")\n",
    "    print(protein)\n",
    "\n",
    "    print(\"\\nNumber of residues:\")\n",
    "    print(len(protein))\n",
    "    \n",
    "    return \"Done\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dd411ed7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accession code:\n",
      ">P04637\n",
      "\n",
      "Name:\n",
      "P53_HUMAN Cellular tumor antigen p53 - Homo sapiens (Human).\n",
      "\n",
      "Protein:\n",
      "MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD\n",
      "\n",
      "Number of residues:\n",
      "393\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Done'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print_protein_info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "046fcb73",
   "metadata": {},
   "source": [
    "# 2\n",
    "\n",
    "Write a subclass called `mCell` that models a mouse cell and extends the `Cell` class defined in blcok 47. It should have self.cytoplasm_size and self.nuclie_size attributes. \n",
    "\n",
    "Code the following method:\n",
    "* .init(*****) constructor. Identify and include all the parameters to ensure that the cell .init() is included.\n",
    "* .count_gene() calculate the ratio of cytoplasm_size and nuclie_size\n",
    "\n",
    "**[Hint: Review the Inheritance section]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2cb3a1a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Cell:\n",
    "    \"\"\" A class representing a cell\"\"\"\n",
    "    \n",
    "    # Constructor \n",
    "    def __init__(self, name):\n",
    "        \"\"\"Input layer\"\"\"\n",
    "        self.name = name\n",
    "        self.organelles = []    # creates a new empty list for each cell\n",
    "        self.genes = []         # same as above\n",
    "    \n",
    "    # The string method\n",
    "    def __str__(self):\n",
    "        return (\"A model of a cell, current instance is {} \".format(self.name))\n",
    "    \n",
    "    # Methods\n",
    "    def add_organelle(self, organelle):\n",
    "        self.organelles.append(organelle)\n",
    "    \n",
    "    def add_gene(self, gene):\n",
    "        self.genes.append(gene)\n",
    "        \n",
    "    def count_organelle(self):\n",
    "        outG = len(self.organelles)\n",
    "        return outG\n",
    "    \n",
    "    def count_gene(self):\n",
    "        outg = len(self.genes)\n",
    "        return outg\n",
    "    \n",
    "    def organelle_gene_ratio(self):\n",
    "        rati = len(self.organelles) / len(self.genes)\n",
    "        return rati\n",
    "        \n",
    "    def view_cell(self):\n",
    "        return self.organelles, self.genes\n",
    "    \n",
    "###\n",
    "\n",
    "class mCell (Cell):\n",
    "    \"\"\"A class extending the cell class\"\"\"\n",
    "    \n",
    "    def __init__(self, name, cytoplasm_size, nuclie_size):\n",
    "        Cell.__init__(self, name)\n",
    "        self.cytoplasm_size = cytoplasm_size\n",
    "        self.nuclie_size = nuclie_size\n",
    "    \n",
    "    def count_gene(self):\n",
    "        return self.cytoplasm_size / self.nuclie_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d758c306",
   "metadata": {},
   "outputs": [],
   "source": [
    "mc1 = mCell(\"mC1\", 300, 30)\n",
    "mc1.count_gene()\n",
    "mc1.add_organelle(\"mcyto\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37e9f548",
   "metadata": {},
   "source": [
    "# 3\n",
    "\n",
    "**Single cell counter class**\n",
    "\n",
    "Write a class called `SCcounter` that represents a single cell counter. It should have a self.count attribute that starts from 0. Also it should have a self.max attribute that represents a limit for the count. And a self.name() that gives a string name to the current instance.\n",
    "\n",
    "Code the following methods:\n",
    "* **.__init__(self, m, name)** constructor. Set self.max to m and self.count to 0\n",
    "* **.increment(self)** increase the count if it is smaller than the max, otherwise complain\n",
    "* **.decrement(self)** decrease the count if it is larger than zero, otherwise complain\n",
    "* **.bulk(self, by)** increase the count by a fixed number if the max allows it others complain\n",
    "* **.reset(self)** reset counter to 0\n",
    "* **.view(self)** view the current max and count\n",
    "* **.__str__(self)** return a string that describes the class and the name of the current instance (this message will be displayed when the object is printed)\n",
    "\n",
    "\n",
    "Write a test program for your counter class. Allocate two objects of type Counter: one called blood with a maximum of 6 and one called kidney with a maximum of 4. \n",
    "\n",
    "Program a while loop to keep asking the user for input. \n",
    "Process input as following:\n",
    "* \"monocyte\" increment blood counter\n",
    "* \"podocyte\" increment kideny counter\n",
    "* \"endothelial\" decrement blood counter\n",
    "* \"macula\" decrement kidney counter\n",
    "* \"error\" reset both counters\n",
    "* \"quit\" exit\n",
    "* Print the .view() method of the two counter objects after each iteration (this will actually call the __str__ method you coded and print the string it returns).\n",
    "\n",
    "**[Hint: Review the OOP section]**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f7f79351",
   "metadata": {},
   "outputs": [],
   "source": [
    "class SCcounter:\n",
    "    \"\"\" A class representing a counter\"\"\"\n",
    "    \n",
    "    # Constructor \n",
    "    def __init__(self, m, name):\n",
    "        \"\"\"Input layer\"\"\"\n",
    "        self.count = 0\n",
    "        self.max = m\n",
    "        self.name = name\n",
    "        \n",
    "    \n",
    "    # Methods\n",
    "    def increment(self):\n",
    "        if self.count + 1 < self.max:\n",
    "            self.count += 1\n",
    "        else:\n",
    "            print(\"The current increase will go above the max\")\n",
    "        return\n",
    "    \n",
    "    def decrement(self):\n",
    "        if self.count > 0:\n",
    "            self.count -= 1\n",
    "        else:\n",
    "            print(\"The current is zero\")\n",
    "        return\n",
    "    \n",
    "    def bulk(self, n=2):\n",
    "        if self.count + n < self.max:\n",
    "            self.count += n\n",
    "        else:\n",
    "            print(\"The current increase will go above the max\")\n",
    "        return\n",
    "\n",
    "    def reset(self):\n",
    "        self.count = 0\n",
    "        return\n",
    "    \n",
    "    def view(self):\n",
    "        return print(\"Max is {} and count is {}\".format(self.max, self.count))\n",
    "    \n",
    "    \n",
    "    def __str__(self):\n",
    "        return (\"A single cell counter machine for {} \".format(self.name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f86c13ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please give your input monocyte\n",
      "Please give your input podocyte\n",
      "Please give your input endothelial\n",
      "Please give your input macula\n",
      "Please give your input quit\n",
      "Non argument\n"
     ]
    }
   ],
   "source": [
    "# Test\n",
    "blood = SCcounter(m=6, name=\"blood\")\n",
    "kidney = SCcounter(m=4, name=\"kidney\")\n",
    "\n",
    "# Process\n",
    "state = \"on\"\n",
    "while state != \"quit\":\n",
    "    state = input(\"Please give your input \")\n",
    "    \n",
    "    if state in(\"monocyte\", \"endothelial\"):\n",
    "        blood.increment()\n",
    "        \n",
    "    elif state in(\"podocyte\", \"macula\"):\n",
    "        kidney.increment()\n",
    "        \n",
    "    elif state in (\"error\"):\n",
    "        blood.reset()\n",
    "        kidney.reset()\n",
    "    else:\n",
    "        print(\"Non argument\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3eede6c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Max is 4 and count is 2\n"
     ]
    }
   ],
   "source": [
    "kidney.view()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9cea445",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"monocyte\" increment blood counter\n",
    "\"podocyte\" increment kideny counter\n",
    "\"endothelial\" decrement blood counter\n",
    "\"macula\" decrement kidney counter\n",
    "\"error\" reset both counters\n",
    "\"quit\" exit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c26dedcf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Max is 6 and count is 2\n",
      "None\n",
      "Max is 4 and count is 2\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(blood.view())\n",
    "print(kidney.view())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
